source/common/formatter/substitution_formatter.cc

```diff
@@ -1918,6 +1925,69 @@ UpstreamHostMetadataFormatter::UpstreamHostMetadataFormatter(const std::string&
                           return host->metadata().get();
                         }) {}
 
+ProxyProtocolTlvsFormatter::ProxyProtocolTlvsFormatter(const std::string& tlv_type) {
+  ASSERT(!tlv_type.empty());
+  // Specified tlv_type must be parsable as an int.
+  try {
+    tlv_type_ = std::stoi(tlv_type);
+  } catch (const EnvoyException& ex) {
+    ENVOY_LOG_MISC(error, "Invalid parameter provided for PROXY_PROTOCOL_TLVS header: {}", tlv_type);
+    // Arbitrary int that won't be used for proxy procotol TLV type. 
+    tlv_type_ = -1; 
+  }
+}
+
+// Parses the proxy protocol TLVs stored in the FilterState with the key "envoy.network.proxy_protocol_options" 
+// and returns a function suitable for accessing the specified metadata from a StreamInfo::StreamInfo. 
+absl::optional<std::string> ProxyProtocolTlvsFormatter::format(const Http::RequestHeaderMap&,
+                                                        const Http::ResponseHeaderMap&,
+                                                        const Http::ResponseTrailerMap&,
+                                                        const StreamInfo::StreamInfo& stream_info,
+                                                        absl::string_view) const {
+  const Envoy::StreamInfo::FilterState& filter_state = stream_info.filterState();
+  auto typed_state = filter_state.getDataReadOnly<Network::ProxyProtocolFilterState>(Network::ProxyProtocolFilterState::key());
+
+  // Value exists but isn't string accessible is a contract violation; throw an error.
+  if (typed_state == nullptr) {
```

Comment:
`typed_state` is a typed pointer. It has the type `Network::ProxyProtocolFilterState*`. I wrote this check to mimic the one performed [here in the original filter state header formatter](https://github.com/10gen/envoy-serverless/blob/aa484fbfcfd6aa19544a860d45e9794ff0f2ee73/source/common/router/header_formatter.cc#L184-L191) after making a call to `getDataReadOnly()`. `getDataReadOnly()` returns `nullptr` if the data does not exist or the data does not match the expected type ([link to generic implementation](https://github.com/10gen/envoy-serverless/blob/aa484fbfcfd6aa19544a860d45e9794ff0f2ee73/envoy/stream_info/filter_state.h#L129-L136)). 

Do you think this check isn't necessary?

Reply:
It would be more appropriate for the comment and log line to reflect that the ProxyProtocolFilterState is not stored in the filter state than how they're currently written. I'll make these changes.